src/hooks/useForm.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useState, useMemo } from 'react';
+
+// Interfaces START
+
+interface Validator {
+  (value: string): string;
+}
+
+interface FormInput {
+  field: string;
+  value: string;
+  validators: Validator[];
+}
+interface ParameterFunction<T> {
+  (value: FormInput): T;
+}
+
+interface Dictionary<T> {
+  [key: string]: T;
+}
+
+// Interfaces END
+
+// Hook START
+const useForm = (settings: FormInput[]) => {
+  const createObjectFrom = (
+    arraySource: FormInput[],
+    keyFunction: ParameterFunction<string>,
+    valueFunction: ParameterFunction<any>
+  ) =>
+    arraySource.reduce((object: Dictionary<any>, item: FormInput) => {
+      object[keyFunction(item)] = valueFunction(item);
+      return object;
+    }, {});
+
+  const calculateErrorList = (
+    value: string,
+    errorFunctions: Validator[]
+  ): string[] => {
+    let errorArray: string[] = [];
+    for (let errorFunction of errorFunctions) {
+      let errorMessage: string = errorFunction(value);
+      if (errorMessage) {
+        errorArray.push(errorMessage);
+      }
+    }
+    return errorArray;
+  };
+
+  const errorFunctionsByKey: Dictionary<[]> = useMemo(() => {
+    return createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => item.validators || []
+    );
+  }, []);
+
+  const [fields, setFields] = useState<Dictionary<string>>(
+    createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => item.value || ''
+    )
+  );
+  const [errors, setErrors] = useState<Dictionary<[]>>(
+    createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => calculateErrorList(item.value, errorFunctionsByKey[item.field])
+    )
+  );
+
+  const disableForm: boolean = useMemo(() => {
+    return !![].concat(...Object.values(errors)).length;
+  }, [errors]);
+
+  const onChange = (event: React.ChangeEvent<{ name: any; value: any }>) => {
+    const { name, value } = event.target;
+
+    if (errorFunctionsByKey[name].length) {
+      const newErrorList: string[] = calculateErrorList(
+        value,
+        errorFunctionsByKey[name]
+      );
+      const oldErrorList: string[] = errors[name];
+
+      let shouldUpdateErrors: boolean =
+        newErrorList.length !== oldErrorList.length;
+
+      if (!shouldUpdateErrors) {
+        for (let i = 0; i < newErrorList.length; i++) {
+          if (newErrorList[i] !== oldErrorList[i]) {
+            shouldUpdateErrors = true;
+            break;
+          }
+        }
+      }
+
+      if (shouldUpdateErrors) {
```

Comment:
If we have a condition to define that errors should not be updated then we should stick with that and only update errors when that condition is set to true. This is how i would solve this but, as always, can be discussed.
Create a function that receives the old and current errors and make every check to see if errors should be updated or not. There not only the length will be checked but also each element inside each array.
const shouldUpdateError = function(oldErrors, newErrors) {
    length comparison and return boolean before full execution
    for cycle here and return boolean
}

Reply:
This comparison is already done by the current code, compare if the sizes are different then errors will be updated O(1).

If sizes are equals then it's going to break the first difference O(n).